airlock/permissions.py

```diff
@@ -0,0 +1,64 @@
+from airlock import exceptions
+from airlock.users import User
+
+
+def check_user_can_view_workspace(user: User | None, workspace_name: str):
+    """
+    This user can access and view the workspace and requests for it.
+    Output checkers can view all workspaces.
+    Authors can view all workspaces they have access to (regardless
+    of the workspace's archive or project ongoing status)
+    """
+    if user is None or (
+        not user.output_checker and workspace_name not in user.workspaces
+    ):
+        raise exceptions.WorkspacePermissionDenied(
+            f"you do not have permission to view {workspace_name} or its requests"
+        )
+
+
+def user_can_view_workspace(user: User | None, workspace_name: str):
+    try:
+        return check_user_can_view_workspace(user, workspace_name) is None
+    except exceptions.WorkspacePermissionDenied:
+        return False
+
+
+def check_user_has_role_on_workspace(user: User | None, workspace_name: str):
+    """
+    This user has an explicit role on the workspace and has permission
+    to create or modify requests.
+    """
+    if user is None or workspace_name not in user.workspaces:
+        raise exceptions.RequestPermissionDenied(
+            f"you do not have permission to author requests for {workspace_name}"
+        )
+
+
+def user_has_role_on_workspace(user: User | None, workspace_name: str):
+    try:
+        return check_user_has_role_on_workspace(user, workspace_name) is None
+    except exceptions.RequestPermissionDenied:
+        return False
```

Comment:
this is fine, although personally I would have been tempted by
```
def user_has_role_on_workspace(user: User | None, workspace_name: str):
    try:
        check_user_has_role_on_workspace(user, workspace_name)
    except exceptions.RequestPermissionDenied:
        return False
    return True
```

Reply:
(I only mention this because it's the beginning of setting a pattern)

Reply:
This is what I did first, then I decided I didn't like the extra `return True`. But if you'd have done it this way too, maybe my first instinct was the right one